### ISSUE TITLE
Improved hashtable lookups

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#1309]
- **Version:** Git Master Branch
- **OS:** Windows

Hash table lookups are rather fast. But when working with hash tables, if the data cache is warm or the tash table has a poor distribution then doing the index lookup can be the majority of the work. The tm_hash interface (in `foundation/hash.inl`) has a wonderful function called `tm_hash_index(table,key)` which isn't used anywhere in the engine (except internally in `hash.inl`), and I think the reason is that there are no `tm_hash_add_using_index()`, `tm_hash_get_using_index()`, `tm_hash_remove_using_index()` etc.

There are a lot of places in the code doing like this:
https://github.com/OurMachinery/themachinery/blob/ea20c67a22819be4ec6292e090fdaac623c9a81e/foundation/the_truth.c#L3630-L3632
https://github.com/OurMachinery/themachinery/blob/ea20c67a22819be4ec6292e090fdaac623c9a81e/foundation/the_truth_assets.c#L1271-L1272
https://github.com/OurMachinery/themachinery/blob/ea20c67a22819be4ec6292e090fdaac623c9a81e/foundation/memory_tracker.c#L309-320
where the index is gotten two to four times instead of once.

This might be seen as a not very important micro optimization, but from personal experience I've seen that since hash tables are used quite a lot it can make a noticable dent in the grand total "performance death by a thousand papercuts".

**Comment by niklas-ourmachinery** [comment]
I agree... I always feel bad when I do more lookups than strictly necessary.

Not 100 % sure that explicitly operating on indices is the right thing to do though... there's a potential for errors if the hash gets modified between the `tm_hash_index()` call and the operations that use the index. Though you could argue that "with great power comes create responsibility".

An alternative approach would be to add functions like:

* `(value, was_added) = tm_get_or_add(h, key, value);`
* `(value, was_removed) = tm_get_and_remove(h, key)`

These functions are not as flexible as the explicit index operations and they also have the unfortunate drawback that returning multiple values is kind of clunky in C (you have to either declare a struct or return one value as a pointer). On the other hand, they don't have the potential for error that the explicit index operations have.

**Comment by Tisten** [comment]
I do like that alternative approach as the mainstream solution. Having more "do multiple things" functions means the code can become safer, more readable(?) and equally fast, so having those in the interface and promoted usage examples would be great.

Some of my examples from above doesn't need any extra functions, like https://github.com/OurMachinery/themachinery/blob/ea20c67a22819be4ec6292e090fdaac623c9a81e/foundation/the_truth_assets.c#L1271-L1272 where doing a `has` before doing an `add` isn't needed. Adding the same key twice isn't a problem. That case might just require education.

The question is if you want to go all-in on the multi-action functions and should remove the `tm_hash_index` from the public interface (i.e rename it to private__), or if there is any case in the code where having the "great power and responsibility" still is the best option. I do think this "great power" interface should be the "advanced special case" and not the "encouraged, first option".

A few other issues with hash tables I saw yesterday:
* sometimes the-machinery doesn't just use calculated hashes as keys, but also raw pointers. This results in a lot of clumping. Since the index calculation is "use the lowest bits" (mixed with the lower part of the upper 32 bits) and bucket count must be power-of-two, pointers tend to end of on 8 or 16 byte positions. The code I looked at was the memory_tracker. Hashing the pointer before using it as a key should help.
* To be able to reason about "do my keys have a good enough distribution or do they clump" it would be really nice to have a development feature (define) to toggle on, which asserts that the clumping of keys are at acceptable levels when resizing the table.
* In the memory tracker I had more than a million keys in each table. I noticed that reindexing it took a while. Not really sure if there is any algoritmic trick to get around this though.

**Comment by niklas-ourmachinery** [comment]
Yes, 8 or 16 byte aligned pointers + power-of-two size hash tables means that 7/8th or 15/16th of the slots will never be mapped to... so you would would get average search lengths of 4 or 8.

But it seems like you saw much greater clustering than that? Clusters of up to 1000 elements, which is a bit surprising. But we should probably pre-hash/mix/swizzle pointer keys anyway.

**Comment by Tisten** [comment]
I also saw that out of the first 100 keys (out of a million) only had 3 keys used, and two of them had been freed again. So I would consider that a cluster of unused keys as well. I guess this will be specific to the system's allocators.

**Comment by niklas-ourmachinery** [comment]
I pushed a fix to reduce clustering in the memory tracker hash tables.
